File: sdk/npm/src/providers/gonebox-email.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "gonebox-email";
const API_BASE = "https://api.gonebox.email/api/v1";

interface CreateInboxResponse {
  success?: boolean;
  data?: {
    id?: string;
    address?: string;
    domain?: string;
    expiresAt?: number;
    ttl?: number;
    existing?: boolean;
  };
}

interface GoneboxMessage {
  id?: string;
  from?: string;
  sender?: string;
  to?: string;
  subject?: string;
  text?: string;
  html?: string;
  date?: string;
  receivedAt?: string;
}

interface MessagesResponse {
  success?: boolean;
  data?: {
    address?: string;
    expiresAt?: number;
    ttl?: number;
    count?: number;
    messages?: GoneboxMessage[];
  };
}

/**
 * 创建 gonebox.email 临时邮箱
 * API: POST /inboxes body: {"domain":"gonebox.email"}
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${API_BASE}/inboxes`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify({ domain: "gonebox.email" }),
  });
  if (!response.ok) {
    throw new Error(`gonebox-email: 创建邮箱失败 ${response.status}`);
  }

  const resp = (await response.json()) as CreateInboxResponse;
  if (!resp.success || !resp.data) {
    throw new Error("gonebox-email: 创建邮箱响应无效");
  }

  const address = String(resp.data.address || "").trim();
  if (!address || !address.includes("@")) {
    throw new Error("gonebox-email: 创建邮箱响应缺少有效地址");
  }

  const expiresAt = resp.data.expiresAt;

  return {
    channel: CHANNEL,
    email: address,
    token: "",
    ...(expiresAt ? { expiresAt } : {}),
  };
}

/**
 * 获取 gonebox.email 邮件列表
 * API: GET /inboxes/{address}/messages
 * @param _token - 未使用（该渠道无需认证）
 * @param email - 邮箱地址
 */
export async function getEmails(
  _token: string,
  email: string,
): Promise<Email[]> {
  const address = String(email || "").trim();
  if (!address) {
    throw new Error("gonebox-email: 邮箱地址为空");
  }

  const response = await fetchWithTimeout(
    `${API_BASE}/inboxes/${encodeURIComponent(address)}/messages`,
    {
      method: "GET",
      headers: { Accept: "application/json" },
    },
  );
  if (!response.ok) {
    throw new Error(`gonebox-email: 获取邮件失败 ${response.status}`);
  }

  const resp = (await response.json()) as MessagesResponse;
  const messages: GoneboxMessage[] = Array.isArray(resp.data?.messages)
    ? resp.data!.messages
    : [];

  return messages.map((msg) => {
    const raw: Record<string, unknown> = {
      id: msg.id || "",
      from: msg.from || msg.sender || "",
      to: msg.to || address,
      subject: msg.subject || "",
      text: msg.text || "",
      html: msg.html || "",
      date: msg.date || msg.receivedAt || "",
    };
    return normalizeEmail(raw, address);
  });
}

File: sdk/rust/src/providers/gonebox_email.rs
/*!
 * gonebox.email — https://gonebox.email
 * 流程：POST /api/v1/inboxes 创建邮箱（无需认证）
 * 收信：GET /api/v1/inboxes/{address}/messages
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::{json, Value};

const API_BASE: &str = "https://api.gonebox.email/api/v1";

/// 创建 gonebox.email 临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let client = http_client();

        let resp = client
            .post(format!("{}/inboxes", API_BASE))
            .header("Content-Type", "application/json")
            .body(r#"{"domain":"gonebox.email"}"#)
            .send()
            .await
            .map_err(|e| format!("gonebox-email: 创建邮箱失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "gonebox-email: 创建邮箱失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("gonebox-email: 解析响应失败: {}", e))?;

        let address = body["data"]["address"]
            .as_str()
            .ok_or("gonebox-email: 响应中未找到邮箱地址")?
            .to_string();

        let expires_at = body["data"]["expiresAt"].as_i64();

        Ok(EmailInfo {
            channel: Channel::GoneboxEmail,
            email: address,
            token: None,
            expires_at,
            created_at: None,
        })
    })
}

/// 获取 gonebox.email 邮件列表
pub fn get_emails(email: &str) -> Result<Vec<Email>, String> {
    let address = email.trim();
    if address.is_empty() {
        return Err("gonebox-email: 邮箱地址为空".into());
    }

    block_on(async {
        let client = http_client();
        let url = format!("{}/inboxes/{}/messages", API_BASE, address);

        let resp = client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("gonebox-email: 获取邮件失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!(
                "gonebox-email: 获取邮件失败 HTTP {}",
                resp.status()
            ));
        }

        let body: Value = resp
            .json()
            .await
            .map_err(|e| format!("gonebox-email: 解析响应失败: {}", e))?;

        let messages = match body["data"]["messages"].as_array() {
            Some(arr) => arr.clone(),
            None => return Ok(Vec::new()),
        };

        if messages.is_empty() {
            return Ok(Vec::new());
        }

        let mut out = Vec::with_capacity(messages.len());
        for msg in &messages {
            let raw = json!({
                "id": msg.get("id"),
                "from": msg.get("from").or_else(|| msg.get("sender")),
                "to": address,
                "subject": msg.get("subject"),
                "date": msg.get("date").or_else(|| msg.get("receivedAt")),
                "html": msg.get("html").or_else(|| msg.get("body")).and_then(|v| v.as_str()).unwrap_or(""),
                "text": msg.get("text").and_then(|v| v.as_str()).unwrap_or(""),
            });
            out.push(normalize_email(&raw, address));
        }
        Ok(out)
    })
}
